File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def ct_same_words(a,b):
    ct = 0
    for i in range(len(a)):
        if a[i] == b[i]:
            ct += 1
    return ct
    

def solution(begin, target, words):
    answer = 0
    
    if target not in words:
        return answer
    
    def bfs(begin, answer):
        q = deque([(begin,0)])
        visited = set([begin])
        ct = 0
        while q:
            v,idx = q.popleft()
            
            if v == target:
                return idx
            
            for word in words:
                if ct_same_words(v, word) == (len(word)-1):
                    if word not in visited:
                        #print(word)
                        visited.add(word)
                        q.append((word, idx+1))
    
    answer = bfs(begin, answer)     
    print(answer)
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    class Words{
        String word;
        int count;
        
        Words(String word, int count){
            this.word = word;
            this.count = count;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        Deque<Words> dq = new ArrayDeque<>();
        boolean[] visited = new boolean[words.length];
        
        dq.offer(new Words(begin, 0));
        while(!dq.isEmpty()){
            Words w = dq.poll();
            
            if(w.word.equals(target)){
                return w.count;            
            }
                    
                
            for(int i=0; i<words.length; i++){
                if(!visited[i] && canTranslate(w.word, words[i])){
                
                    //System.out.println(cnt + " " + words[i]);
                    dq.offer(new Words(words[i], w.count + 1));
                    visited[i] = true;
                }
                
            }
        }
        
        return 0;
    }
    
    public boolean canTranslate(String a, String b){
        int tmp = 0;
        for(int i=0; i<a.length(); i++){
            if(a.charAt(i) != b.charAt(i)){
                tmp++;
            }
            
            if(tmp > 1){
            return false;
            }
        }
        
        return true;
    }
}